### PR TITLE
Redis Initial Code

### PR DESCRIPTION
**Pull request by adivik2000**
This PR has the necessary code for Redis Vector Database. For some reason, I'm unable to get any results from querying and I need some help figuring out what might be the reason. I followed this doc here -> https://github.com/redis/redis-py/blob/master/docs/examples/search_vector_similarity_examples.ipynb

My result: 
```
{
  "result": {
    "total": 0,
    "duration": 2.492189407348633,
    "docs": []
  }
}
```
The duration suggests that the query is not malformed etc. I'll leave it to you from here.

### CODE REVIEW

File: vectordb.py [hunk 1]
```diff
@@ -102,3 +107,100 @@ def delete_index(self) -> str:
 #     ],
 #     "namespace": "",
 # }
+
+
+class RedisDB(VectorDatabase):
+    def __init__(self, index_name):
+        super().__init__(index_name)
+        self.batch_size = 50
```
Comment (NirantK) [inline vectordb.py:115]: Let's not declare this twice!

File: vectordb.py [hunk 2]
```diff
@@ -102,3 +107,100 @@ def delete_index(self) -> str:
...
+                    "metadata": metadata,
+                },
+            )
+
+            # Write batch
+            if i % self.batch_size == 0:
+                pipeline.execute()
+
+        # Cleanup final batch
+        pipeline.execute()
```
Comment (NirantK) [inline vectordb.py:181]: What does `pipeline` do?

File: vectordb.py [hunk 3]
```diff
@@ -102,3 +107,100 @@ def delete_index(self) -> str:
...
+    def query(self, query_embedding: List[float]) -> dict:
+        # base_query = (
+        #     f"(@metadata:{ Old Testament })=>[KNN {self.top_k} @vector $query_embedding AS score]"
+        # )
+        # logger.info(f"base query: {base_query}")
+        query = (
+            Query("*=>[KNN 3 @vector $query_embedding AS score]")
+            .return_fields("text", "score")
+            .dialect(2)
+        )
```
Comment (NirantK) [inline vectordb.py:194]: What does `dialect(2)` do?

File: vectordb.py [hunk 4]
```diff
@@ -102,3 +107,100 @@ def delete_index(self) -> str:
...
+            # Constants
+            schema = (
+                TextField(name="text"),
+                TextField(name="metadata"),
+                VectorField(
+                    "embedding",
+                    "FLAT",
+                    {
+                        "TYPE": "FLOAT32",
+                        "DIM": 768,
```
Comment (NirantK) [inline vectordb.py:143]: Parameterize 768!

File: vectordb.py [hunk 5]
```diff
@@ -14,7 +19,7 @@ def __init__(self, index_name, top_k: int = 3):
         # Load the dataset
         self.dataset = load_dataset(
             "Cohere/wikipedia-22-12-simple-embeddings", split="train"
-        )
+        ).select(range(1000))
```
Comment (NirantK) [inline vectordb.py:22]: This 1K should be a parameter perhaps?

File: vectordb.py [hunk 6]
```diff
@@ -102,3 +107,100 @@ def delete_index(self) -> str:
...
+                mapping={
+                    "text": text,
+                    "vector": np.array(embedding, dtype=np.float32).tobytes(),
+                    "metadata": metadata,
+                },
+            )
+
+            # Write batch
+            if i % self.batch_size == 0:
+                pipeline.execute()
```
Comment (NirantK) [inline vectordb.py:178]: Why is this a sequential loop? Should be a multithread/multi-processing perhaps given that Redis can take it unlike Pinecone?

**Review by NirantK — CHANGES_REQUESTED** [review]

File: vectordb.py (hunk 3 above)
Comment (adivik2000) [reply vectordb.py:194]: From: https://redis.io/docs/stack/search/reference/vectors/

Screenshot 2023-05-10 at 10 07 58 PM

File: vectordb.py (hunk 4 above)
Comment (adivik2000) [reply vectordb.py:143]: Hey! I've already done this on other PR. Just waiting for the merges to finish and I'll clean them all up. Don't worry about this.

File: vectordb.py (hunk 5 above)
Comment (adivik2000) [reply vectordb.py:22]: Will do.

File: vectordb.py (hunk 6 above)
Comment (adivik2000) [reply vectordb.py:178]: Will try that. I took it from langchain -> https://github.com/hwchase17/langchain/blob/3637d6da6e51cd8b96a9a4f91eb71e8db21a8b20/langchain/vectorstores/redis.py#L211

File: vectordb.py (hunk 2 above)
Comment (adivik2000) [reply vectordb.py:181]: https://buildatscale.tech/what-is-redis-pipeline/

File: vectordb.py (hunk 1 above)
Comment (adivik2000) [reply vectordb.py]: Took care of this.

**Pull request merged**